File: python_app/models/history.py
"""History Record model - translated from HISTREC.cpy copybook.

Mirrors the COBOL HISTORY-RECORD structure with composite key
(portfolio_id, date, time, seq_no) and history data fields.
"""

from enum import StrEnum

from pydantic import BaseModel, Field


class HistoryRecordType(StrEnum):
    """History record type codes from 88-level values in HISTREC."""

    PORTFOLIO = "PT"
    POSITION = "PS"
    TRANSACTION = "TR"


class HistoryActionCode(StrEnum):
    """History action codes from 88-level values in HISTREC."""

    ADD = "A"
    CHANGE = "C"
    DELETE = "D"


class HistoryRecord(BaseModel):
    """Full history record translated from COBOL HISTORY-RECORD.

    Maps to HISTREC.cpy copybook fields:
    - HIST-KEY (composite key)
    - HIST-DATA (history details with before/after images)
    """

    # Key fields (HIST-KEY)
    portfolio_id: str = Field(max_length=8, description="HIST-PORTFOLIO-ID")
    date: str = Field(max_length=8, description="HIST-DATE: YYYYMMDD")
    time: str = Field(max_length=6, description="HIST-TIME: HHMMSS")
    seq_no: str = Field(max_length=4, description="HIST-SEQ-NO")

    # Data fields (HIST-DATA)
    record_type: HistoryRecordType = Field(description="HIST-RECORD-TYPE: PT/PS/TR")
    action_code: HistoryActionCode = Field(description="HIST-ACTION-CODE: A/C/D")
    before_image: str = Field(default="", max_length=400, description="HIST-BEFORE-IMAGE")
    after_image: str = Field(default="", max_length=400, description="HIST-AFTER-IMAGE")
    reason_code: str = Field(default="", max_length=4, description="HIST-REASON-CODE")

    @property
    def composite_key(self) -> str:
        """Build the composite key matching VSAM KSDS key structure."""
        return f"{self.portfolio_id}{self.date}{self.time}{self.seq_no}"

File: shared/src/types/history.ts
/**
 * History record types derived from COBOL copybook HISTREC.cpy (lines 6-27)
 */

export type RecordType = 'PORT' | 'POS' | 'TRN';

export type ActionCode = 'ADD' | 'CHANGE' | 'DELETE';

export interface HistoryRecord {
  portfolioId: string;     // HIST-PORTFOLIO-ID (8 chars)
  date: string;            // HIST-DATE (ISO 8601)
  time: string;            // HIST-TIME (HH:MM:SS)
  seqNo: string;           // HIST-SEQ-NO (4 chars)
  recordType: RecordType;  // HIST-REC-TYPE PORT/POS/TRN
  actionCode: ActionCode;  // HIST-ACTION-CODE ADD/CHANGE/DELETE
  beforeImage: string;     // HIST-BEFORE-IMAGE (400 chars)
  afterImage: string;      // HIST-AFTER-IMAGE (400 chars)
  reasonCode: string;      // HIST-REASON-CODE (4 chars)
  processDate: string;     // HIST-PROCESS-DATE (ISO 8601)
  processUser: string;     // HIST-PROCESS-USER (8 chars)
}
